numpy/core/getlimits.py

```diff
@@ -29,32 +29,95 @@ def _fr1(a):
         a.shape = ()
     return a
 
+
 class MachArLike:
     """ Object to simulate MachAr instance """
-
-    def __init__(self,
-                 ftype,
-                 *, eps, epsneg, huge, tiny, ibeta, **kwargs):
-        params = _MACHAR_PARAMS[ftype]
-        float_conv = lambda v: array([v], ftype)
-        float_to_float = lambda v : _fr1(float_conv(v))
-        float_to_str = lambda v: (params['fmt'] % array(_fr0(v)[0], ftype))
-
-        self.title = params['title']
+    def __init__(self, ftype, *, eps, epsneg, huge, tiny, smallest_normal,
+                 ibeta, smallest_subnormal=None, **kwargs):
+        self.params = _MACHAR_PARAMS[ftype]
+        self.ftype = ftype
+        self.title = self.params['title']
+        self._smallest_subnormal = smallest_subnormal
         # Parameter types same as for discovered MachAr object.
-        self.epsilon = self.eps = float_to_float(eps)
-        self.epsneg = float_to_float(epsneg)
-        self.xmax = self.huge = float_to_float(huge)
-        self.xmin = self.tiny = float_to_float(tiny)
-        self.ibeta = params['itype'](ibeta)
+        self.epsilon = self.eps = self.float_to_float(eps)
+        self.epsneg = self.float_to_float(epsneg)
+        self.xmax = self.huge = self.float_to_float(huge)
+        self.xmin = self.tiny = self.float_to_float(tiny)
+        self.ibeta = self.params['itype'](ibeta)
+        self.smallest_normal = self.float_to_float(smallest_normal)
+        self._str_smallest_normal = self.float_to_str(self.smallest_normal)
         self.__dict__.update(kwargs)
         self.precision = int(-log10(self.eps))
-        self.resolution = float_to_float(float_conv(10) ** (-self.precision))
-        self._str_eps = float_to_str(self.eps)
-        self._str_epsneg = float_to_str(self.epsneg)
-        self._str_xmin = float_to_str(self.xmin)
-        self._str_xmax = float_to_str(self.xmax)
-        self._str_resolution = float_to_str(self.resolution)
+        self.resolution = self.float_to_float(
+            self.float_conv(10) ** (-self.precision))
+        self._str_eps = self.float_to_str(self.eps)
+        self._str_epsneg = self.float_to_str(self.epsneg)
+        self._str_xmin = self.float_to_str(self.xmin)
+        self._str_xmax = self.float_to_str(self.xmax)
+        self._str_resolution = self.float_to_str(self.resolution)
+
+    @property
+    def smallest_subnormal(self):
+        """Return the value for the smallest subnormal.
+
+        Returns
+        -------
+        smallest_subnormal : float
+            value for the smallest subnormal
+
+        Warns
+        -----
+        UserWarning
+            If the calculated value for the smallest subnormal is zero.
+        """
+        if not self._smallest_subnormal:
+            value = nextafter(self.ftype(0), self.ftype(1), dtype=self.ftype)
+        else:
+            value = self._smallest_subnormal
```

Comment:
Lets move this chunk into the `__init__`?  The rest would have to be here I guess.  Hopefully, it just can't be triggered in any case.

(Maybe we should add a comment that it should not be possible to trigger the warning, so if anyone wants to remove it at some point, they are less confused then us.)